## Understanding chromatin biology using high throughput sequencing


| Audience | Computational Skills | Prerequisites | Duration |
:----------|:----------|:----------|:----------|
| Biologists | Intermediate | None | Shell basics|


### Learning Objectives

1.	Understand the necessity for, and use of, the command line interface (bash) and HPC for analyzing high-throughput sequencing data.
2.	Understand best practices for designing a ChIP-seq / CUT&RUN / ATAC-seq experiment.
3.	Perform the steps involved in going from raw FASTQ files to peak calls for an individual sample.
4.	Review qualitative ways to assess peak calls and if they support the hypothesis


> These materials were developed for a trainer-led workshop, but are also amenable to self-guided learning.

### Description

This repository has teaching materials for the Understanding chromatin biology using high throughput sequencing workshop. 

> Pre-requisite: Introduction to Shell

This workshop focuses on teaching computational skills to enable the effective use of an high-performance computing environment to implement a ChIP-seq data analysis workflow. In addition to running the workflow from FASTQ files to peak calls, the workshop covers best practice guidelines for ChIP-seq experimental design and data organization/management and data visualization for quality control.

* Introduction to ChIP-seq, CUT&RUN and ATAC-seq
* Experimental design considerations
* QC using FASTQC
* Alignment theory and considerations for ChIP-seq
* Alignment and filtering of reads
* Peak calling
* Peak visualization (using deepTools)
  

---

### Citation

To cite material from this course in your publications, please use:

> Meeta Mistry, Jihe Liu, Mary Piper, William Gammerdinger, & Radhika Khetani. (2022, January 6). hbctraining/Intro-to-ChIPseq-flipped: ChIP-seq Data Analysis Lessons from HCBC (first release). Zenodo. https://doi.org/10.5281/zenodo.5825871

A lot of time and effort went into the preparation of these materials. Citations help us understand the needs of the community, gain recognition for our work, and attract further funding to support our teaching activities. Thank you for citing this material if it helped you in your data analysis.

---

*These materials have been developed by members of the teaching team at the [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/). These are open access materials distributed under the terms of the [Creative Commons Attribution license](https://creativecommons.org/licenses/by/4.0/) (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.*

* *Some materials used in these lessons were derived from work that is Copyright Â© Data Carpentry (http://datacarpentry.org/). 
All Data Carpentry instructional material is made available under the [Creative Commons Attribution license](https://creativecommons.org/licenses/by/4.0/) (CC BY 4.0).*

